From 044472db32a0282de23c28697a7dec9bc66fee23 Mon Sep 17 00:00:00 2001
From: Mike Achtelik <mike.achtelik@gmail.com>
Date: Tue, 28 Jun 2022 09:55:25 +0200
Subject: Android A11Y: Fix deadlock in
 QtAndroidAccessibility::runInObjectContext()

On android the event loop is normally blocked, when the application
is suspended, e.g. when it enters the background or when the screen is
locked (see android.app.background_running). This leads to a problem
when we try to process events after this happens, e.g. when android
sends us an ACTION_CLEAR_ACCESSIBILITY_FOCUS event after the event loop
is suspended. While handling it we eventually call
QtAndroidAccessibility::runInObjectContext() which tries to do a
blocking call on the object context, however, with the event loop being
suspended we run into a deadlock which leads to an ANR.  So we need to
make sure to never make a blocking call while the event loop is
suspended.

Task-number: QTBUG-102594
Pick-to: 6.4 6.3 6.2 5.15
Change-Id: I33f0440a3da84fb4bdae5ab0fc10d514c73f23ad
---
 src/plugins/platforms/android/androidjniaccessibility.cpp | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git x/qtbase/src/plugins/platforms/android/androidjniaccessibility.cpp y/qtbase/src/plugins/platforms/android/androidjniaccessibility.cpp
index 1043d0058d..66218e3d1f 100644
--- x/qtbase/src/plugins/platforms/android/androidjniaccessibility.cpp
+++ y/qtbase/src/plugins/platforms/android/androidjniaccessibility.cpp
@@ -94,7 +94,13 @@ namespace QtAndroidAccessibility
     template <typename Func, typename Ret>
     void runInObjectContext(QObject *context, Func &&func, Ret *retVal)
     {
-        QMetaObject::invokeMethod(context, func, Qt::BlockingQueuedConnection, retVal);
+        if (!QtAndroid::blockEventLoopsWhenSuspended()
+            || QGuiApplication::applicationState() != Qt::ApplicationSuspended) {
+            QMetaObject::invokeMethod(context, func, Qt::BlockingQueuedConnection, retVal);
+        } else {
+            __android_log_print(ANDROID_LOG_WARN, m_qtTag,
+                                "Could not run accessibility call in object context, event loop suspended.");
+        }
     }
 
     void initialize()
-- 
2.32.1 (Apple Git-133)
